### PR TITLE
feat: #2447 CONSEP API update germination test (header + activity)

### LINKED ISSUE #2447: CONSEP API – Update Germination Test (Header + Activity)
**Describe the task**
Update test status (accept/complete), dates, rank, and activity data.

**Acceptance Criteria**
- [ ] Cannot accept a test that has not been marked as complete
- [ ] Begin Date is mandatory when: accepted checked, or complete checked, or comment entered,  or Test End entered
- [ ] Test end must be after begin date
- [ ] Backend determines original_test_ind and current_test_ind; client cannot set directly
- [ ] If Original or Current updated, backend must reset flags on other tests for same seedlot, category, and activity type.
- [ ] Test update + activity update + post-update processing must commit or rollback as a unit.

**Additional context**
Database transactions and validations are documented here: https://apps.nrs.gov.bc.ca/int/confluence/display/FSADT2/Germination+Testing

### PR DESCRIPTION
**Pull request by xiaopeng0202**
## Summary

Adds `PATCH /api/germination-tests/{riaKey}` (oracle-api) that updates the germination test header (`CNS_T_TSC_TEST_RESULT`) and its activity (`CNS_T_RQST_ITM_ACTVTY`) atomically, per the Confluence "Germination Testing" spec ("Update (test and activity)" + "Post update processing").

Closes #2447

## Acceptance criteria → tests

| AC | Test |
|---|---|
| Cannot accept a test not marked complete | `updateGerminationTest_acceptWithoutComplete_throwsUnprocessable` |
| Begin Date mandatory when accepted/complete/comment/Test End | `updateGerminationTest_completeWithoutBeginDate_throwsUnprocessable`, `…_commentWithoutBeginDate_…` |
| Test end must be after begin date | `updateGerminationTest_endNotAfterBegin_throwsUnprocessable` |
| Backend determines original/current_test_ind; client cannot set | flags absent from `GerminationTestUpdateFormDto`; computed in service (`updateGerminationTest_accepted*` flag-matrix tests) |
| Original/Current updated → reset flags on sibling tests | `updateGerminationTest_acceptedNoSiblings_setsOriginalCurrentAndRankA`, `…_acceptedLatestEnd_replacesCurrentRankA` |
| Test + activity + post-update commit/rollback as a unit | single `@Transactional` service method; optimistic-lock conflict tests (`…_staleTestResultTimestamp_throwsConflict`, `…_staleActivityTimestamp_throwsConflict`) |

Also implemented from spec: Test End defaults to now when Complete checked; first-time Complete removes `CNS_T_ASSIGND_GERM_LOC` row; Related Test Results columns zeroed on save; rank recomputation (STD + Accepted); RTS/TST first-completed-activity commitment processing; revised start/end derivation.

## API contract notes (FE ticket impact)

Optimistic locking implemented per spec SQL: request must include `testResultUpdateTimestamp` and `riaUpdateTimestamp` (values read from the GET header endpoint). Stale timestamp → **409 Conflict**. Other statuses: 400 invalid body, 404 unknown riaKey, 422 business-rule violation.

## Deviations from Confluence pseudocode (please confirm)

1. `If Test End < qry.min_end → updOriginal = 0` is a no-op contradicting the field-description table (original = -1 when no earlier-ending test). **Implemented as -1.**
2. `revised_end_dt` CASE uses `:TestEnd + duration` in the TestEnd-IS-NULL branch (always NULL); field description says Begin Date + duration. **Implemented as begin + duration.**
3. Commitment processing is unconditional in the pseudocode, but its comment says "the first **completed** activity processes the commitment". **Implemented to run only when Complete is checked** — reviewer please confirm.
4. Germ Tolerance transaction on Complete is out of scope (belongs to the replicates/counts ticket).
5. (post-Copilot-review) Flag comparisons follow the field-description table rather than the pseudocode's strict `>`/`<`: original/current stay set on ties (covers re-saving the same test), and the sibling min/max queries treat indicator columns as non-zero-true since this repo's own endpoints write `1` while legacy data uses `-1`.

## Test plan

- [x] `mvn verify` green (all oracle-api tests + checkstyle)
- [x] 16 new service tests (validations, flag matrix, 409s, commitment/germ-loc) + 5 new MockMvc endpoint tests (200/400/404/409/422)
- [ ] Manual smoke against dev once a FE consumer exists

---

Thanks for the PR!

Deployments, as required, will be available below:
- [Frontend](https://nr-spar-4.apps.silver.devops.gov.bc.ca/)
- [Backend](https://nr-spar-2504-backend.apps.silver.devops.gov.bc.ca/swagger-ui/index.html)
- [Oracle-API](https://nr-spar-2504-oracle-api.apps.silver.devops.gov.bc.ca/actuator/health)

Please create PRs in draft mode.  Mark as ready to enable:
- [Analysis Workflow](https://github.com/bcgov/nr-spar/actions/workflows/analysis.yml)

After merge, new images are deployed in:
- [Merge Workflow](https://github.com/bcgov/nr-spar/actions/workflows/merge.yml)

### CODE REVIEW

File: oracle-api/src/main/java/ca/bc/gov/oracleapi/service/consep/TestResultService.java [hunk 1]
```diff
@@ -753,6 +758,239 @@ private void validateReplicateTotals(
...
+      LocalDateTime minEnd = testResultRepository
+          .findMinCompletedAcceptedEndDate(seedlotNumber, activityType, category);
+      LocalDateTime maxEnd = testResultRepository
+          .findMaxCompletedAcceptedEndDate(seedlotNumber, activityType, category);
+      if (maxEnd == null || effectiveEnd.isAfter(maxEnd)) {
+        updCurrent = -1;
+      }
+      if (minEnd == null || effectiveEnd.isBefore(minEnd)) {
+        updOriginal = -1;
+      }
```
[the remark this answers is not in the document]
Comment (xiaopeng0202) [reply oracle-api/src/main/java/ca/bc/gov/oracleapi/service/consep/TestResultService.java]: Fixed in 2e1acd41. Switched to the field-description semantics: current/original stay set unless a sibling ends strictly later/earlier (`!effectiveEnd.isBefore(maxEnd)` / `!effectiveEnd.isAfter(minEnd)`), so ties — including re-saving the same test — keep the flags. Added `updateGerminationTest_acceptedEndEqualsSiblingEnds_keepsFlagsSet` to pin it.

File: oracle-api/src/main/java/ca/bc/gov/oracleapi/service/consep/TestResultService.java [hunk 2]
```diff
@@ -753,6 +758,239 @@ private void validateReplicateTotals(
...
+    boolean complete = Boolean.TRUE.equals(dto.testCompleteInd());
+
+    validateGerminationTestUpdate(dto, storedTest, accept, complete);
+
+    // Spec: when Complete is checked, Test End defaults to now.
+    LocalDateTime effectiveEnd = dto.actualEndDateTime();
+    if (complete && effectiveEnd == null) {
+      effectiveEnd = LocalDateTime.now();
+    }
+
```
[the remark this answers is not in the document]
Comment (xiaopeng0202) [reply oracle-api/src/main/java/ca/bc/gov/oracleapi/service/consep/TestResultService.java]: Fixed in 2e1acd41. `effectiveEnd` is now resolved before validation and the end-after-begin rule checks `effectiveEnd`, so the defaulted value is covered too.

File: oracle-api/src/main/java/ca/bc/gov/oracleapi/service/consep/TestResultService.java [hunk 3]
```diff
@@ -753,6 +758,239 @@ private void validateReplicateTotals(
...
+          "Seed withdrawal date cannot be in the future");
+    }
+
+    boolean storedComplete = storedTest.getTestCompleteInd() != null
+        && storedTest.getTestCompleteInd() != 0;
+    if (storedComplete
+        && !dto.testCategoryCode().equals(storedTest.getTestCategory())) {
+      throw new ResponseStatusException(HttpStatus.UNPROCESSABLE_ENTITY,
+          "Category cannot be updated once the test is complete");
+    }
```
[the remark this answers is not in the document]
Comment (xiaopeng0202) [reply oracle-api/src/main/java/ca/bc/gov/oracleapi/service/consep/TestResultService.java:997]: No NPE is possible in this direction: `testCategoryCode` is `@NotNull` (bean-validated at the endpoint) and it is the receiver — `nonNull.equals(null)` returns false, which is also the desired outcome (a null stored category counts as a change and is rejected). Leaving as-is.

File: oracle-api/src/main/java/ca/bc/gov/oracleapi/repository/consep/TestResultRepository.java [hunk 4]
```diff
@@ -206,4 +207,145 @@ long countAcceptedStdRankA(
...
+  @Query("""
+      SELECT MIN(a.actualEndDateTime)
+      FROM TestResultEntity t
+      JOIN ActivityEntity a
+        ON a.riaKey = t.riaKey
+      WHERE a.seedlotNumber = :seedlotNumber
+        AND t.activityType = :activityType
+        AND t.testCategory = :testCategory
+        AND t.testCompleteInd = -1
+        AND t.acceptResult = -1
```
[the remark this answers is not in the document]
Comment (xiaopeng0202) [reply oracle-api/src/main/java/ca/bc/gov/oracleapi/repository/consep/TestResultRepository.java]: Fixed in 2e1acd41 — both MIN and MAX queries now use `<> 0`. This matters in this repo specifically: `updateTestResultStatusToCompleted`/`...Accepted` write 1 while legacy CONSEP data uses -1, so `= -1` would have missed rows completed through the existing endpoints. The `anotherTestIsCurrent` check in the service was aligned to non-zero as well.

File: oracle-api/src/main/java/ca/bc/gov/oracleapi/repository/consep/TestResultRepository.java [hunk 5]
```diff
@@ -206,4 +207,145 @@ long countAcceptedStdRankA(
...
+  @Query("""
+      SELECT MAX(a.actualEndDateTime)
+      FROM TestResultEntity t
+      JOIN ActivityEntity a
+        ON a.riaKey = t.riaKey
+      WHERE a.seedlotNumber = :seedlotNumber
+        AND t.activityType = :activityType
+        AND t.testCategory = :testCategory
+        AND t.testCompleteInd = -1
+        AND t.acceptResult = -1
```
[the remark this answers is not in the document]
Comment (xiaopeng0202) [reply oracle-api/src/main/java/ca/bc/gov/oracleapi/repository/consep/TestResultRepository.java]: Fixed in 2e1acd41 together with the MIN query (see reply there).

File: oracle-api/src/main/java/ca/bc/gov/oracleapi/dto/consep/GerminationTestUpdateFormDto.java [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+    @Schema(
+        description =
+            "update_timestamp of the test result row as read (optimistic lock)"
+    )
+    @NotNull
+    LocalDateTime testResultUpdateTimestamp,
+    @JsonFormat(shape = JsonFormat.Shape.STRING, pattern = "yyyy-MM-dd'T'HH:mm:ss")
+    @Schema(description = "update_timestamp of the activity row as read (optimistic lock)")
+    @NotNull
+    LocalDateTime riaUpdateTimestamp
```
[the remark this answers is not in the document]
Comment (xiaopeng0202) [reply oracle-api/src/main/java/ca/bc/gov/oracleapi/dto/consep/GerminationTestUpdateFormDto.java:56]: Fixed in 2e1acd41. `GerminationTestHeaderDto` now exposes `testResultUpdateTimestamp` and `riaUpdateTimestamp` (both in the GET response and in the PATCH response body), so clients can read-modify-write with the optimistic-lock values.

File: oracle-api/src/main/java/ca/bc/gov/oracleapi/repository/consep/ActivityRepository.java [hunk 7]
```diff
@@ -111,4 +111,33 @@ int updateTimestampWhereMatch(
...
+      """)
+  int updateGerminationTestActivity(
+      @Param("riaKey") BigDecimal riaKey,
+      @Param("actualBeginDateTime") LocalDateTime actualBeginDateTime,
+      @Param("actualEndDateTime") LocalDateTime actualEndDateTime,
+      @Param("revisedStartDate") LocalDate revisedStartDate,
+      @Param("revisedEndDate") LocalDate revisedEndDate,
+      @Param("riaComment") String riaComment,
+      @Param("intermediateCleaner") Integer intermediateCleaner,
+      @Param("expectedUpdateTimestamp") LocalDateTime expectedUpdateTimestamp);
```
[the remark this answers is not in the document]
Comment (xiaopeng0202) [reply oracle-api/src/main/java/ca/bc/gov/oracleapi/repository/consep/ActivityRepository.java:146]: Fixed in 2e1acd41. Parameter renamed to `commentIsCritical`; the JPQL still sets `a.intermediateCleaner` and the javadoc now notes the INTRMDT_CLEANR_IND column is repurposed for the critical-comment flag.

**Review by mightycox — APPROVED** [review]
